qutebrowser/browser/bookmarks.py

```diff
@@ -0,0 +1,140 @@
+    removed = pyqtSignal(str)
+
+    def __init__(self, parent=None):
+        """Initialize and read bookmarks."""
+        super().__init__(parent)
+
+        self.bookmarks = collections.OrderedDict()
+
+        if standarddir.config() is None:
+            self._lineparser = None
+        else:
+            bookmarks_directory = os.path.join(standarddir.config(),
+                                               'bookmarks')
+            if not os.path.isdir(bookmarks_directory):
+                os.makedirs(bookmarks_directory)
+            self._lineparser = lineparser.LineParser(
+                standarddir.config(), 'bookmarks/urls', parent=self)
+            for line in self._lineparser:
+                if not line.strip():
+                    # Ignore empty or whitespace-only lines.
+                    continue
+                try:
+                    url, title = line.split(maxsplit=1)
+                except ValueError:
+                    message.error(0, "Invalid bookmark '{}'".format(line))
+                else:
+                    self.bookmarks[url] = title
+            filename = os.path.join(standarddir.config(), 'bookmarks/urls')
+            objreg.get('save-manager').add_saveable(
+                'bookmark-manager', self.save, self.changed,
+                filename=filename)
+
+    def save(self):
+        """Save the bookmarks to disk."""
+        if self._lineparser is not None:
+            self._lineparser.data = [' '.join(tpl)
+                                     for tpl in self.bookmarks.items()]
+            self._lineparser.save()
+
+    def bookmark_add(self, win_id, url, title):
+        """Add a new bookmark.
+
+        Args:
+            win_id: The window ID to display the errors in.
+            url: The url to add as bookmark.
+            title: The title for the new bookmark.
+        """
+        if not url.isValid():
+            urlutils.invalid_url_error(win_id, url, "save quickmark")
+            return
+        urlstr = url.toString(QUrl.RemovePassword | QUrl.FullyEncoded)
+
+        # We don't raise cmdexc.CommandError here as this can be called async
+        # via prompt_save.
+        if not title:
+            message.error(win_id, "Can't set mark with empty title!")
+            return
+        if not urlstr:
+            message.error(win_id, "Can't set mark with empty URL!")
+            return
+
+        if urlstr in self.bookmarks:
+            message.error(win_id, "Bookmark already exists!")
+        else:
+            self.bookmarks[urlstr] = title
+            self.changed.emit()
+            self.added.emit(title, urlstr)
+            message.info(win_id, "Bookmark added")
+
+    def bookmark_del(self, url):
+        """Delete a bookmark.
+
+        Args:
+            url: The url of the bookmark to delete.
+        """
+        try:
+            del self.bookmarks[url]
+        except KeyError:
+            raise cmdexc.CommandError("Bookmark '{}' not found!".format(url))
```

Comment:
Since this isn't a command, it probably shouldn't raise a `cmdexc.CommandError`. The `KeyError` here should never happen as it's only called via the completion anyways, right? Then I'd say remove the whole try/except-block.